TrueCacheAPI.md

```diff
@@ -0,0 +1,236 @@
+# True Cache API
+
+This design document defines a true Cache API for use cases such as Hibernate 2L cache.
+Throughout the document we will use Hibernate 2L as use case for explaining rationale behind the API.
+
+
+## Read-Only Cache
+The most basic of cache APIs, supporting only read-only data, should allow:
+
+1. Check if data is present in cache, if present return it, otherwise fetch it from source and store it into the cache.  
+This complex operation is the essence of a caching API.
+In Hibernate, this is split into two Hibernate callbacks into the cache region api: 
+`CachedDomainDataAccess.get` and (maybe) `CachedDomainDataAccess.putFromLoad`.
+
+2. Invalidate individual or all cached entries.
+The user might want to invalidate individual or all cache entries in the cache. 
+For example, if a fresher data has been stored into the database via methods other than Hibernate. 
+
+4. Provide an estimate of the number of entries in cache.
+A user might want to find out the estimated number of entries in a cache.
+This is useful for making assertions about the cached contents.
+
+Here's an an asynchronous API that would match these requirements:
+
+```java
+interface ReadOnlyCache<K, V> {
+  
+  /**
+   * Checks if the cache contains a value associated with the key.
+   * If the value is present, it's returned immediately wrapped in a CompletionStage.
+   * 
+   * If the value is not present, the function is called asynchronously to fetch the value from an unknown source.
+   * If the supplier returned a value, the value is cached and then returned to the caller wrapped in a CompletionStage.
+   * 
+   * If the supplier didn't find a value, null is returned to the caller wrapped in a CompletionStage.
+   *  
+   * @param key to look up in the cache
+   * @param fetcher called if not value is associated with the key in the cache
+   * @return a future containing the cached entry if present, otherwise the value returned by the supplier 
+   */
+  CompletionStage<V> get(K key, Function<K, V> fetcher);
```

Comment:
From an asynchronous perspective, should `Function<K, V>` be `Function<K, CompletionStage<V>` instead?

Reply:
👍

Reply:
We could provide even a blocking and non blocking variant.

Reply:
Do you want to redeem `null` or use `Optional`?

Reply:
> We could provide even a blocking and non blocking variant.

Not sure about providing a blocking variant since it can be derived from the non blocking one? A blocking one can be created out of taking `V` and converting it to `CompletionStage<V>`, but I see that more like a static helper method or something. I want to avoid overloading.

Reply:
> Do you want to redeem null or use Optional?

Eventually should be Optional. So this would eventually look like this?

```
CompletionStage<Optional<V>> get(K key, Function<K, CompletionStage<Optional<V>> fetcher);
```